agent/xds/listeners.go

```diff
@@ -1068,14 +1057,64 @@ func (s *ResourceGenerator) makeTerminatingGatewayListener(
 		}
 	}
 
+	for _, svc := range cfgSnap.TerminatingGateway.ValidDestinations() {
+		clusterName := connect.ServiceSNI(svc.Name, "", svc.NamespaceOrDefault(), svc.PartitionOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+
+		intentions := cfgSnap.TerminatingGateway.Intentions[svc]
+		svcConfig := cfgSnap.TerminatingGateway.ServiceConfigs[svc]
+
+		cfg, err := ParseProxyConfig(svcConfig.ProxyConfig)
+		if err != nil {
+			// Don't hard fail on a config typo, just warn. The parse func returns
+			// default config if there is an error so it's safe to continue.
+			s.Logger.Warn(
+				"failed to parse Connect.Proxy.Config for linked destination",
+				"destination", svc.String(),
+				"error", err,
+			)
+		}
+
+		var dest *structs.DestinationConfig
+		if cfgSnap.TerminatingGateway.DestinationServices[svc].ServiceKind == structs.GatewayServiceKindDestination {
+			dest = &svcConfig.Destination
+		} else {
+			return nil, fmt.Errorf("invalid gateway service for destination %s", svc.Name)
+		}
+		clusterChain, err := s.makeFilterChainTerminatingGateway(cfgSnap, clusterName, svc, intentions, cfg.Protocol, dest)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make filter chain for cluster %q: %v", clusterName, err)
+		}
+		l.FilterChains = append(l.FilterChains, clusterChain)
+	}
+
 	// Before we add the fallback, sort these chains by the matched name. All
 	// of these filter chains are independent, but envoy requires them to be in
 	// some order. If we put them in a random order then every xDS iteration
 	// envoy will force the listener to be replaced. Sorting these has no
 	// effect on how they operate, but it does mean that we won't churn
 	// listeners at idle.
 	sort.Slice(l.FilterChains, func(i, j int) bool {
-		return l.FilterChains[i].FilterChainMatch.ServerNames[0] < l.FilterChains[j].FilterChainMatch.ServerNames[0]
+		si := ""
+		sj := ""
+		if len(l.FilterChains[i].FilterChainMatch.PrefixRanges) > 0 {
+			si += l.FilterChains[i].FilterChainMatch.PrefixRanges[0].AddressPrefix +
+				"/" + l.FilterChains[i].FilterChainMatch.PrefixRanges[0].PrefixLen.String() +
+				":" + l.FilterChains[i].FilterChainMatch.DestinationPort.String()
```

Comment:
I don't think this pattern works if AddressPrefix is ipv6. Need to do something like [::1]:8500 to be correct.

Reply:
This is an interesting callout because we added this work thinking we would support CIDR blocks, but since you can't pass this information from the connect proxy to the gateway, it wasn't going to work. Since even IP addresses have a fake SNI as `server_name`, turns out we don't even need any changes to this method.